Fix ci and benchmarks

Sankara-Jefferson opened the pull request:
This PR addresses several critical areas focusing on CI reliability, infrastructure improvements, and code quality enhancements. The changes fall into four main categories:

**Original CI Failures:**

- Tests were randomly failing due to port conflicts between concurrent database instances
- Multiple test suites trying to use port 27017 for MongoDB simultaneously
- MySQL and PostgreSQL tests colliding on their default ports
- These conflicts caused connection timeouts and "address already in use" errors
- The CI builds were failing due to a combination of static analysis errors and inconsistent code formatting, which also impacted overall code maintainability.
- Previous: Manual documentation led to code/doc drift

**Root Cause Analysis:**

- IPDX workflows were starting database services without configurable ports
- Our custom tests were also starting databases on default ports
- When both ran together, or when parallel tests ran, they would conflict
- No proper cleanup between test runs meant ports could stay occupied

**Problem Impact:**

- CI builds were becoming unreliable and flaky
- PRs required multiple retries and still failed
- Test failures weren't actual code issues but infrastructure problems


**Implementation** 
**CI/Testing Infrastructure**

- Improved CI workflow configuration with better caching strategies
- Added MongoDB integration test support and documentation
- Implemented cache cleanup workflow
- Enhanced test stability and reliability

**Code Quality & Architecture**

- Refactored client code structure and API organization
- Improved thread safety in analytics package
- Enhanced error handling and type safety
- Fixed duration handling in client configuration

**Documentation**

- Added integration test requirements documentation
- Improved API documentation
- Reorganized documentation generation
- Restored original docgen.sh functionality and regenerated complete CLI documentation with proper environment variable handling

**Dependencies**

- Updated multiple dependencies to newer versions
- Removed unused dependencies

**Note**: I will address the risks and Mitigations in this [issue](https://github.com/data-preservation-programs/singularity/issues/582) . Kindly add anything missing on the issue as a comment to the PR for reference. I will lift them to the issue when creating a new pr for them.

lanzafame reviewed (approved):
If ci passes, I guess this is good. This size pr is ridiculous though and in general singularity PRS need to become more manageable

ianconsolata reviewed (changes requested):
Can you please document what the error is that you were trying to fix, and why so many change are needed to accomplish that goal? This looks like a complete overhaul of the CI system, including removing the CI tooling the the IPDX team maintains for ecosystem golang projects. I'd prefer not to depart from using the ecosystem standard CI approach without good reason.

In particular, it looks like some change were made to the docsgen workflow, and all the docs were deleted from the repo. Was that intended? Why do you feel like the docs should no longer be checked into the codebase?

Sankara-Jefferson commented:
@ianconsolata @lanzafame – Sorry about the large diff. The actual code changes are minimal; most of the changes (41,190 deletions vs. 586 insertions) are due to documentation cleanup and file reorganization within the client/swagger directory.
I have updated the pr's description to document the issues it addresses. 

**IPDX Removal Rationale**
The removal of IPDX was initially aimed at resolving CI failures. However, during the analysis, I identified several limitations in IPDX that would hinder future scalability:

- Database Service Configuration: No service health checks, inability to run multiple DB versions, fixed environment variables, limited cleanup mechanisms.
- Test Environment Management: Lack of proper resource isolation and cleanup.
- Test Suite Organization: Soon there will be need for dynamic port allocation, isolated test environments, custom configurations, and better test categorization.

**As a next steps**
If you think we should keep it, I can reintroduce IPDX in a follow-up PR with enhancements that address these gaps, including:
- Configurable DB ports.
- Proper test environment isolation.
- Clear separation between IPDX workflows and custom test workflows.
- Improved resource management.
This PR focuses on stabilizing CI while laying the groundwork for these improvements.

Sankara-Jefferson commented:
@ianconsolata I've restored docgen.sh and updated the description to that effect. All tests passed and the CI pipeline has passed all checks. If possible, let me know your thoughts. I'd like it merged to unblock Anjor working over the weekend.

ianconsolata commented on .github/actions/go-test-setup/action.yml:
```diff
@@ -3,14 +3,21 @@ description: Setup Cache
+      shell: bash
+      run: |
+        rm -rf ~/.cache/go-build
+        rm -rf ~/go/pkg/mod
```
Also, why isn't this needed on the go-check-setup action?

ianconsolata commented on .github/workflows/go-check.yml:
```diff
@@ -14,19 +15,88 @@ concurrency:
+        run: mysql -h127.0.0.1 -P3306 -usingularity -psingularity -e "SELECT VERSION();"
+
+      - name: Verify PostgreSQL connection
+        run: PGPASSWORD=singularity psql -h localhost -U singularity -d singularity -c "SELECT version();"
```
Why do you create a MySQL connection and a PSQL connection? Singularity seems to support both, so I could understand wanting to test with both, but I don't see you passing a connection string to Singularity anywhere so I think this test is still using sqllite. Have I missed where you configure the database you set up for testing?

https://data-programs.gitbook.io/singularity/installation/deploy-to-production

parkan commented on .github/workflows/go-check.yml:
```diff
@@ -14,19 +15,88 @@ concurrency:
+        run: go build ./...
+
+      - name: Run tests
+        run: go test -v ./...
```
why are you running go test inside go check?

parkan commented:
while I'm at it, I am considering deprecating/removing mongodb code as it adds a ton of unnecessary complexity in testing and confuses introspecting the codebase... this is a separate concern but it was brought to my attention due to the daemon management code here
